hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/cache/CacheManager.java

```diff
@@ -0,0 +1,312 @@
+            if (log.isTraceEnabled()) {
+                publishCache.entrySet().stream().forEach(e -> log.tracef("Published: %s", e.getValue()));
+            }
+
+            // This will include group trigger conditions, which don't need to be published but it's
+            // not easy to efficiently filter them out.
+            Collection<Condition> conditions = definitions.getAllConditions();
+            final Set<CacheKey> activeKeys = new HashSet<>();
+            for (Condition c : conditions) {
+                // external conditions are evaluated by the relevant external alerter. Data is not metrics-based
+                // so don't waste energy adding it to the publishing cache.
+                if (c instanceof ExternalCondition) {
+                    continue;
+                }
+
+                DataIdKey dataIdKey = new DataIdKey(c.getTenantId(), c.getDataId());
+                CacheKey cacheKey = dataIdKey.getCacheKey();
+                if (dataIdKey.isValid() && !activeKeys.contains(cacheKey)) {
+                    activeKeys.add(cacheKey);
+                    if (!currentlyPublished.contains(cacheKey)) {
+                        publish(dataIdKey);
+                    }
+                }
+                if (c instanceof CompareCondition) {
+                    String data2Id = ((CompareCondition) c).getData2Id();
+                    DataIdKey dataIdKey2 = new DataIdKey(c.getTenantId(), data2Id);
+                    CacheKey cacheKey2 = dataIdKey2.getCacheKey();
+                    if (dataIdKey2.isValid() && !activeKeys.contains(cacheKey2)) {
+                        activeKeys.add(cacheKey2);
+                        if (!currentlyPublished.contains(cacheKey2)) {
+                            publish(dataIdKey2);
+                        }
+                    }
+                }
+            }
+
+            final Set<CacheKey> doomedKeys = new HashSet<>();
+            if (!currentlyPublished.isEmpty()) {
+                currentlyPublished.stream()
+                .filter(k -> !activeKeys.contains(k))
+                .forEach(k -> doomedKeys.add(k));
+            }
+            unpublish(doomedKeys);
+
+            log.debugf("Published after update=%s", publishCache.size());
+            if (log.isTraceEnabled()) {
+                publishCache.entrySet().stream().forEach(e -> log.tracef("Published: %s", e.getValue()));
+            }
+        } catch (Exception e) {
+            log.error("FAILED to load conditions to create Id filters. All data being forwarded to alerting!", e);
+            return;
+        }
+    }
+
+    private void publish(DataIdKey dataIdKey) {
+        CacheKey cacheKey = dataIdKey.getCacheKey();
+        if (cacheKey != null && publishCache != null) {
+            // This logic assumes that alerting is the only writer for the shared publishCache
+            log.debugf("Publishing:%s ", cacheKey);
+            publishCache.put(cacheKey, dataIdKey.getDataId());
```

Comment:
You are right about the classloader.  Because we are deployed as an EAR this works fine for the new metrics distribution.  And that is why I chose to avoid using Strings and stay with this CacheKey class that I already had in place when you discovered the classloader issue.  It could be a problem for other clients that want to filter but aren't packaged as an EAR.  But then again, we may be moving the filtering into alerting, in the sendData processing.

And you are right again, one of the ideas here was to remove Metrics dependencies in Alerting and have only the reverse, let Metrics depend on Alerting.  So don't use Hm's MetricId class as a key managed by Ha.